### ISSUE TITLE
GOSS Proofing

### ISSUE THREAD

**Issue by aceson28** [microsoft/LightGBM#4388]
is there any other way to make proof that ordinary gain is the same as gain in GOSS other than using O()? because my basic skill from statistics and I don't understand when the proof using O() 
any mathematical way to prove that?

**Comment by shiyu1994** [comment]
Hi @aceson28 Could you please refer to the proof in the Appendix of LightGBM paper?
https://papers.nips.cc/paper/2017/hash/6449f44a102fde848669bdd9eb6b76fa-Abstract.html
In Section 1 of the Appendix (`1 Theorem 3.2 and its proof`), there's no big-O notation contained in the proof.

**Comment by shiyu1994** [comment]
if you have any further problem, please feel free to post here. Sorry for the late response.

**Comment by aceson28** [comment]
@shiyu1994  i saw that papers but there are no appendix

**Comment by aceson28** [comment]
Thanks @shiyu1994 i will check it and understand it
how about EFB? is there are step by step algorithms in a word or a video not in pseudo code

**Comment by shiyu1994** [comment]
I think the pseudo code and the source code in this repo is the best material for EFB. Do you have any problem about it?

**Comment by aceson28** [comment]
TBH yes, I don't familiar with pseudocode, and C language too
because my basic skill from statistics and my first computation language is R and then I learn Python
I was searching other journals about LightGBM but there are no one explained about EFB, even LightGBM step by step is none
Btw can you correct me?
LightGBM steps that i know is:

1. first we make a histogram bins for every variable (The boundaries of histogram is fix for a whole process)
 in this  process data sampled and divided into m numbers of bin with almost the same number of data
for numerical data, we sort it and divide it into a bins (almost the same number of data for every bins)
for categorical data, every factor or class of variable has a bin and we sort it by (sum of gradient/sum of hessian)
this sampled data used just for making a histogram boundaries
in this process, missing data will be ignored
2. Do EFB:
We defined K as a max conflict where K=number of data(real data)/10000
count a conflict between 2 variable and make a matrix of conflict
conflict is a condition where variable 1 and variable 2 is not a zero
we make graph coloring with variable as edge/circle and number of conflict between 2 variable as line and we sort sum of conflict for every variable
we start cuting a line of variable with a variable which has the highest number of conflicts 
the line will be cut if number of conflict <= K
If V1 (Variable 1) has a connection line with V2, V2 has a connection line with V3, but V1 doesn't has a connection line with V3 so a connection between V2 and V3 will be cut
after that process we will have a couple of variables to be merge, number of variables to be marge into a single variable can be more than 2
The merging process will be adding an offset with the max value of the first variable
supposedly variable 1 has a value [0,10] and variable 2 has a value[0,20]
so for variable 1 who in variable 2 has 0 value, the new value is a same value
for variable 2 who in variable 1 has 0 value, the new value will be +10 than before
for a condition where variable 1 and variable 2 have a 0 value, the new value will be 0
My question from this section is:
 a. is there are labeling process before this process, if there is no labeling process where does the value come from?  the value in this process is a real value or bin number?
 b. how about missing value?
 c. in the merging process how about if there are more than 2 variables? is V3 will be added +30?
 d. In the merging process how we merge a conflict?

3. GOSS
 we use a% of top gradient and sampled b*(1-a)% of the other gradient for every iteration

4. Prediction of Nodes
In GBDT the leaf values will be the average of the gradient for regression and (sum of gradient /sum of hessian) for classification
My question is the value of nodes will be using real data or sampled data in GOSS?

can you correct me or give me some more information about LightGBM Process

**Comment by shiyu1994** [comment]
Hi @aceson28,

For histogram construction, it is basically correct.

For the EFB part, the way you described to bundle features into groups is not accurate.
1. First, we build a graph, in which each vertex represents a feature. And whenever there're conflicts between two features, we add an edge between the two corresponding vertices in the graph.
2. Then, we sort the features by descending order or the degrees of the vertices.
3. Initialize the set of feature groups as an empty set.
4. Traverse the features in the order from step 2. For each feature, we find the feature group in the feature group set with the least number of conflicts (the number of conflicts of a feature with a feature group is the sum of all conflicts with features in that group). If the minimum number of conflicts is smaller than `K` (which is number of data / 10000), then we find a feasible group to add the feature into. Otherwise we build a new group with that single feature, and add the new group into the feature group set.
5. The merging process is basically the same as you described.
For your questions:
a. The values are bin numbers. (But I did not understand what's called a "labeling process", could you explain more?).
b. Missing values is already mapped to some bin (usually the last bin, if `zero_as_missing=false`, otherwise the bin that 0 is mapped to).
c. Yes. V3 will be added by 30.
d. For now, the bin value of the data instance with conflicted items, is the last item among them being processed. This will depends on the detailed input format (CSV or LibSVM), but in most cases it will be the last feature among these items in the original data.
e. We don't consider the boundaries of a merged feature group, we only consider the boundaries of single features. Note that feature groups are only for histogram construction which is for training, but boundaries are useful in inference.
Note that for efficiency, the actual implementation may be a little different with the process above. But this is the basic idea. For the full details, you are encouraged to read the source code.

For GOSS part, we select `a * n` (`a` is in `(0,1)`) instances with the largest absolute values of gradient values (which is weighted by hessian values, actually). Then we randomly sample `b * n` (`b` is in (0,1)) instances from the rest `n - a * n` instances. gradient and hessian values of these randomly sampled instances are multiplied by `(1-a) / b` for unbiased estimation.

For prediction of nodes, 
>In GBDT the leaf values will be the average of the gradient for regression
Actually, only for the default `l2` regression, there are many other regressions whose hessian values are not constant.
The value of leaves is calculated only with sampled instances in GOSS and any other sampling strategy (for example, the ordinary bagging).

Please feel free to post any further questions here.

**Comment by aceson28** [comment]
I think the result of grouping feature that I know and that you explained from step 1 - 4 it will give the same result, am I right?
for my question:
 a. labeling that i mean is label encoder, so the first bin will be 0 the second one is 1 and etc.
 b. if the missing value will be mapped to the last bin,so in the tree, missing value will always go to "No" Branch?
 d. TBH, i don't understand anything about what your answered for this number
 e. if the boundaries of the merged features are not considered, then what is the function of merged features for histogram construction?

**Comment by shiyu1994** [comment]
@aceson28 Sorry for the late response. For the EFB feature grouping process, I suggest to leave out the graph stuff for better understanding, so I wrote the process in 1~4. I'm sorry, but I don't quite understand the operations you described in the graph. I think we can describe it with a clearer language with the feature group sets etc. 

a. The labeling process for the original categorical values is before the feature grouping. In other words, we would map the categorical values into continuous integer values, e.g., [0, 2, 5] --> [0, 1, 2].
b and c. Whether the missing value goes to left or right branch is determined by the split gain. If going the the left branch results in a higher gain, the missing value will be grouped into the left branch, otherwise it is n the right branch.
d. The key message is that if there's a conflict happen, what's in the bin value of the conflict item is data dependent. And we can not make any conclusion about it.
e. Note that the feature grouping is done with the discretized version of feature values (bin values), so in this step the boundaries of the single feature histogram is not related, since the values are already discretized.

Please feel free to post any further questions here.

**Comment by aceson28** [comment]
ok I get it for some question
for a. labeling I mean is the label of discretization bins not for categorical.
b. so what is the function of mapped missing value to last/first bin?
d. can I say if there are conflicts, the new value is same as the first variable value?
e. if it is like that, I don't see the effect of merged variable

**Comment by shiyu1994** [comment]
b. Mapping missing values to the last/first bin makes skipping them easier when finding the best split threshold.
d. No, when there are conflicts, the final value in the merged feature bin is undefined.
e. Merging features allowing us to leave out the missing values and save memory for sparse features, and handles the merged feature as a dense feature, which is more cache friendly than sparse features with only non-missing values being stored.

**Comment by aceson28** [comment]
b and e. is after merging the variable and find the best candidate for split point, the histogram will not use ?
d. can you refer what the algorithm do in code? in what lines?

**Comment by shiyu1994** [comment]
@aceson28 Sorry, could you please elaborate more on your first question?
For corresponding code of the behavior in d, please refer to following code,
https://github.com/microsoft/LightGBM/blob/b1c261a5d3653669e709e5f1a5bc5d17ee31a114/include/LightGBM/feature_group.h#L192-L213
and
https://github.com/microsoft/LightGBM/blob/b1c261a5d3653669e709e5f1a5bc5d17ee31a114/src/io/dense_bin.hpp#L69-L82
As we can see, when pushing a feature value into a group (to be accurate, a non-multi value feature group, which is the case that the paper of LightGBM shows), it simply stores the bin value into the corresponding line position in the merged feature group. So if feature A and feature B have a conflict in line _i_, then if feature A of line _i_ is pushed first, feature B of line _i_ will overwrite the value of feature A in line _i_, otherwise feature A of line _i_ will overwrite the value of feature B in line _i_.

**Comment by skullknight15** [comment]
Thank you @shiyu1994  for your explanations.
Indeed @aceson28 , there are almost no explanations about EFD. Even the paper provides really superficial details about it. I am working currently with LGBM and I am stuck trying to  understand the EFD part. When we evaluate the conflicts between features, do we evaluate conflicts between original features or bins of the histograms ? If we evaluate conflicts between bins of histograms, then the minimum conflict ratio is number of bins/10000 ? I'm really confused.
Let's say now that we are evaluating conflicts between the original continuous features. Let us consider a dataset composed of 100000 samples * D features. If minimum number of allowed conflict is k= len(data)/10000 = 10 for instance, and my features are not sparse, which is exactly my case, there is absolutely not a single zero over all features over all samples. What happens in this case, all features are in conflict ? or there is a functions that analyzes first if the features are sparse.  I understand the conflicts computation and the features bundling, but I cab't see if conflicts are deduced by analyzing teh bins or the real feature values. Any explanation is welcome

**Comment by jameslamb** [comment]
reopening this discussion so it can be seen by others who can answer @skullknight15 's question.

cc @shiyu1994 @tongwu-msft @hzy46

**Comment by shiyu1994** [comment]
Hi @skullknight15. Thanks for using LightGBM.
The conflict is calculated after the features are discretized into bins.
And if all features are dense, all features will be conflict with each other. And each single feature will be a single feature group.
Feel free to post any further questions here.

**Comment by shiyu1994** [comment]
Hi @skullknight15,
For your questions:
1. We merge with discretized bin values because it allows us to better align the merging process with subsequent histogram trick. (bin values are the number of the bin where an original feature value falls into its histogram)
2. Yes the code snippet describes the merging well. The only thing to note is, in the latest version of LightGBM, sometimes we use `the most frequent bin of a feature` as the `missing value` in feature bundling, to maximize the speedup. So all the `!=0` and `==0` should be replaced by `!=most_freq_bin` and `==most_freq_bin`. You may see the following code for reference. From how the `sparse_rate_` is calculated, you may see which values are defined as `missing` for EFB in each case.
https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/src/io/bin.cpp#L504-L519
Note that the paper can only describes the basic idea of EFB, because there're too many implementation details.

**Comment by skullknight15** [comment]
@shiyu1994 thank you for the explanation. It makes more sense to me now.
So depending on the condition (most_freq_bin_ != default_bin_ && max_sparse_rate < kSparseThreshold), LGBM either uses 0 as or most_freq_bin_ as  for deducing conflicts. Can you tell me how kSparseThreshold is computed?
I have one last question. When we merge the original features f1 with feature f2, we add max(f1) to feature f2 according to the conditions described in my last comment. Is it possible to saturate the feature values of f2 when for example for sample i, f2[i] + max(f1) > max(f2) or it doesn't matter since after merging we map the merged features into bin values to construct a histogram ?

**Comment by shiyu1994** [comment]
@skullknight15 kSparseThreshold  is a constant, which is defined in include/LightGBM/bin.h.
>Is it possible to saturate the feature values of f2
It doesn't matter at all. After merging the features, we still guarantee that the bin values and histograms are correctly aligned.

**Comment by skullknight15** [comment]
@shiyu1994 Okay.
So if I resume the whole process for EFB and GOSS. 

- **Preprocessing step using EFB:**
1- The K original features are mapped into bins.
2- The Binned K features are ordered based on their conflicts (deduced  using the count of most_freq_bin_ as missing values).
3- Starting from the most conflicted feature to the least one, we seek to merge it with another feature(s) based on the  conflict ratio specified by len(data)/10000
4- If feature f1 and feature f2 can be merged, we merge their original values f1[i] and f2[i] at each sample index i based on their bin values F1[i] and F2[i] at index i as:
- feature_bundle12[i] = f2[i]+ max(f1) if F1[i]==most_freq_bin and F2[i] !=most_freq_bin
- feature_bundle12[i] = f1[i] if (F1[i]!=most_freq_bin and F2[i] !=most_freq_bin) or (F1[i]==most_freq_bin and F2[i] ==most_freq_bin) or (F1[i]!=most_freq_bin and F2[i] ==most_freq_bin)
5- After constructing the feature bundles, lets say we have now L bundles where L << K, each feature bundle l will be re-mapped into bin values (can you confirm this) in order to perform gradient boosting and GOSS

- **Gradient boosting and GOSS on EFB data:**
1- We have now the data set of size **len(data)*L** , GOSS will try now to reduce len(data) based on the gradient values to new_data, where len(new_data)<< len(data). 
Here, the gradient boosting and GOSS processes are performed on bin values of the histograms since it is much faster to compute split points based on bins rather than on original feature values.  **Note**: GOSS is enabled at the 20-th iteration as stated in the hyper-parameters documentation of LGBM.
2- So from the 20-th to last iteration, the histogram building costs are reduced from _O_(len(data)*K) to _O_(len(new_data)*L), where new_data << data, and L<<K.

Can you confirm this ? If it's looks correct to you, then you can close this issue.
Thank you again for your time and consideration

**Comment by shiyu1994** [comment]
Yes, your description is correct with the high-level ideas. But still, it is encouraged to read the code to fully know the details, especially for EFB. Because there're some simplifications or approximations to the original idea to make it more efficient. I'll close this issue for now, and if you have any further questions, please feel free to reopen it. Thanks.

**Comment by skullknight15** [comment]
@shiyu1994 thank you for the explanations.

**Comment by MAxx8371** [comment]
@shiyu1994 
In skullknight15's statement, for a sample _**i**_, if F1[i]!=most_freq_bin and F2[i] !=most_freq_bin, it means sample _**i**_ is a conflict and feature_bundle12[i]== f1[i]. <br/>
But in your previous response, it says that the bin value of the data instance with conflicted items, is the last item among them being processed. Does that means that if we processed F1[i] first, then the feature_bundle12[i] should be F2[i]. Am I wrong?
<br/>
Thank you in advance for your response!

**Comment by shiyu1994** [comment]
@MAxx8371 Yes, your statement is accurate. EFB does causes some loss of information. That's why we try to reduce conflicts in bundling.

**Comment by MAxx8371** [comment]
@shiyu1994 Well, I understand. Thank you for your response. Appreciate it.
